Correctly track Surface object

arminn opened the pull request:
With patch c6fde3bad5 to the WlBuffer was
added logic which may call clear() method on
the Surface object. The main issue is there
are several WlBuffer objects and only one Surface.
Since raw ptr was used, memory issues may appear.
he current code adds weak_pt(not shared_pt - since
we don't want to change the overcomplicated
destruction process).

Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn committed:
Correctly track Surface object

With patch c6fde3bad5 to the WlBuffer was
added logic which may call clear() method on
the Surface object. The main issue is there
are several WlBuffer objects and only one Surface.
Since raw ptr was used, memory issues may appear.
he current code adds weak_pt(not shared_pt - since
we don't want to change the overcomplicated
destruction process).

Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

al1img commented on src/displayBackend/wayland/Connector.cpp:
```diff
@@ -76,6 +77,8 @@ void Connector::pageFlip(FrameBufferPtr frameBuffer, FlipCallback cbk)
 	}
 
 	mSurface->draw(frameBuffer, cbk);
+	WlBuffer* buffer = dynamic_cast<WlBuffer*>(frameBuffer.get());
```
No need for new var: `dynamic_cast<WlBuffer*>(frameBuffer.get())->setSurface(mSurface);`

arminn replied:
Done

al1img commented on src/displayBackend/wayland/FrameBuffer.cpp:
```diff
@@ -41,18 +41,17 @@ WlBuffer::WlBuffer(DisplayBufferPtr displayBuffer,
-	if (mSurface)
-	{
-		mSurface->clear();
+	auto surfaceShared = mSurface.lock();
```
separate var definition from other code by empty line. `surfaceShared` sounds ambiguous: just `surface` or `surfacePtr` or `surfaceSharedPtr` would be better.

arminn replied:
Done

al1img commented on src/displayBackend/wayland/FrameBuffer.cpp:
```diff
@@ -95,14 +94,13 @@ void WlBuffer::onRelease()
+void WlBuffer::setSurface(SurfacePtr surface)
 {
 	lock_guard<mutex> lock(mMutex);
-
```
Keep this empty line.

arminn replied:
Done

al1img commented on src/displayBackend/wayland/FrameBuffer.hpp:
```diff
@@ -73,7 +73,7 @@ class WlBuffer : public DisplayItf::FrameBuffer
 	}
 
-	void setSurface(Surface* surface);
+	void setSurface(SurfacePtr surface);
```
Probably better to have weak ptr here as a parameter to emphasize that it will not take ownership but only temporary using?

arminn replied:
Done

al1img commented on src/displayBackend/wayland/Surface.cpp:
```diff
@@ -212,6 +212,16 @@ void Surface::init(wl_compositor* compositor)
 
 void Surface::release()
 {
+	{
```
If there is a race, I don't see how these changes may help. Changing the order of destroying callback will not solve the issue IMHO.

arminn replied:
Patch replaced

arminn committed:
Destroy DisplayCommandHandler correctly

The Wayland callbacks (which eventually lead
to DisplayCommandHandler::sendFlipEvent) start
from Surface and used Connector and EventRingBuffer
objects. Prior to Connector object destruction,
all Wayland callbacks should be disabled.

Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

al1img commented on src/displayBackend/wayland/FrameBuffer.cpp:
```diff
@@ -41,18 +41,17 @@ WlBuffer::WlBuffer(DisplayBufferPtr displayBuffer,
-	{
-		mSurface->clear();
+	if (surfaceSharedPtr) {
+		surfaceSharedPtr.get()->clear();
```
no need to get raw pinter to access clear method: `surfaceSharedPtr->clear()` should work.

arminn replied:
Done

al1img commented on src/displayBackend/DisplayCommandHandler.cpp:
```diff
@@ -104,6 +104,7 @@ DisplayCommandHandler::~DisplayCommandHandler()
 	LOG(mLog, DEBUG) << "Delete command handler, connector name: "
 					 << mConnector->getName();
 
+	mConnector->stopProcessing();
```
1. Commit message: add space `callbacks (`
2. I believe such an issue should be resolved by deleting/creating object order: I don't see someone else owning the connector. If it is the case, stop processing functionality can be called in connector destructor.

arminn replied:
Done

al1img commented on src/displayBackend/wayland/Connector.cpp:
```diff
@@ -80,6 +80,15 @@ void Connector::pageFlip(FrameBufferPtr frameBuffer, FlipCallback cbk)
 
+void Connector::stopProcessing()
+{
+	if (mSurface)
```
The same here. Only this connecter is owning the surface. If so, disable callback can be called in surface destructor.

arminn replied:
Done

arminn commented:
This is proper solution, but close it until "proper" time.
